src/proposal.mligo

```diff
@@ -112,12 +112,25 @@ let check_proposal_limit_reached (config, propose_params, store : config * propo
     then (failwith("MAX_PROPOSALS_REACHED") : storage)
     else store
 
-let freeze_on_ledger (tokens, addr, ledger, total_supply, unfrozen_token_id, frozen_token_id : nat * address * ledger * total_supply * token_id * token_id)
-    : (ledger * total_supply) =
-  let (ledger, total_supply) = debit_from (tokens, addr, unfrozen_token_id, ledger, total_supply) in
+let freeze_on_ledger (tokens, addr, ledger, total_supply, frozen_token_id, governance_token : nat * address * ledger * total_supply * token_id * governance_token)
+    : (operation * ledger * total_supply) =
+  // Call transfer on token_contract to transfer `token` number of
+  // tokens from `addr` to the address of this contract.
+  let
+    transfer_param : transfer_params =
+      [{ from_ = addr; txs = [ { to_ = Tezos.self_address; token_id = governance_token.token_id; amount = tokens } ] }] in
+  let token_contract =
+    begin
+      match (Tezos.get_entrypoint_opt "%transfer" governance_token.address : ((transfer_params contract) option)) with
+        | Some (c) -> c
+        | None -> (failwith "Token contract address failed typecheck" : (transfer_params contract))
```

Comment:
Please convert this error (and its occurrence in `unfreeze_on_ledger`) to the same form as the other errors and add documentation, e.g. as with `NOT_ENOUGH_FROZEN_TOKENS`.

Reply:
Have renamed the error to `BAD_TOKEN_CONTRACT`.